napari/_qt/qt_range_slider_popup.py

```diff
@@ -0,0 +1,168 @@
+        width = self.fm.boundingRect('8' * len(text)).width() + 4
+        width = max(self.min_width, min(width, self.max_width))
+        if width > self.min_width:
+            # don't ever make the label smaller ... it causes visual jitter
+            self.min_width = width
+        self.setFixedWidth(width)
+
+    def update_position(self):
+        x = self.get_pos() - self.width() / 2
+        y = self.slider.handle_radius + 6
+        self.move(QPoint(x, -y) + self.slider.pos())
+
+    def mouseDoubleClickEvent(self, event):
+        self.selectAll()
+
+
+class QRangeSliderPopup(QtPopup):
+    def __init__(self, parent=None, horizontal=True, precision=0, **kwargs):
+        """A popup window that contains a range slider and linked LineEdits.
+
+        Parameters
+        ----------
+        parent : QWidget, optional
+            Will like be an instance of QtLayerControls.  Note, providing
+            parent can be useful to inherit stylesheets.
+        horizontal : bool, optional
+            Whether the slider is oriented horizontally, by default True.
+            (Vertical orientation has not been tested much)
+        precision : int, optional
+            Number of decimal values in the labels, by default 0
+        **kwargs
+            all additional keyword arguments will be passed to the RangeSlider
+        """
+        super().__init__(parent)
+        self.precision = precision
+        layout = QHBoxLayout()
+        self.slider = (
+            QHRangeSlider(parent=parent, **kwargs)
+            if horizontal
+            else QVRangeSlider(parent=parent, **kwargs)
+        )
+        self.slider.setMinimumHeight(18)
+        self.frame.setLayout(layout)
+        self.frame.setContentsMargins(0, 8, 0, 0)
+        cmin, cmax = self.slider.values()
+        get_min_pos = partial(getattr, self.slider, 'display_min')
+        get_max_pos = partial(getattr, self.slider, 'display_max')
+        self.curmin_label = LabelEdit(self._numformat(cmin), self, get_min_pos)
+        self.curmax_label = LabelEdit(self._numformat(cmax), self, get_max_pos)
+        rmin, rmax = self.slider.range
+        self.range_min_label = LabelEdit(self._numformat(rmin))
+        self.range_max_label = LabelEdit(self._numformat(rmax))
+        self.range_min_label.setAlignment(Qt.AlignRight)
+        layout.addWidget(self.range_min_label)
+        layout.addWidget(self.slider, 50)
+        layout.addWidget(self.range_max_label)
+        self.layout = layout
+        self.slider.setFocus()
+        self.curmin_label.editingFinished.connect(self._curmin_label_changed)
+        self.curmax_label.editingFinished.connect(self._curmax_label_changed)
+        self.range_min_label.editingFinished.connect(self._range_label_changed)
+        self.range_max_label.editingFinished.connect(self._range_label_changed)
+        self.slider.valuesChanged.connect(self._on_values_change)
+        self.slider.rangeChanged.connect(self._on_range_change)
```

Comment:
This giant block of text is hard to follow... Perhaps break it up into commented sections?